device/pcie/pci_device.cpp

```diff
@@ -174,59 +207,38 @@ inline void memcpy_from_device(void *dest, const void *src, std::size_t num_byte
     return device_ids;
 }
 
-PCIDevice::PCIDevice(int device_id, int logical_device_id) {
-    // TODO: use C++ constructor to do everything
-    // TODO: make public member vars const
-    // TODO: get logical_id out of here
-    this->device_id = device_id;
-    this->logical_id = logical_device_id;
-    setup_device();
-
-    this->info = read_device_info(device_fd);
-
-}
-
-PCIDevice::~PCIDevice() {
-    close_device();
-}
-
-
-void PCIDevice::setup_device() {
-    this->device_fd = find_device(this->device_id);
-    this->numa_node = determine_numa_node(this->device_fd);
-    this->pcie_device_id = get_pcie_info(this->device_id, "pcie_device_id");
-    this->pcie_revision_id = get_pcie_info(this->device_id, "revision");
-    this->arch = detect_arch(pcie_device_id, pcie_revision_id);
-    this->architecture_implementation = tt::umd::architecture_implementation::create(static_cast<tt::umd::architecture>(arch));
-
+PCIDevice::PCIDevice(int device_number, int logical_device_id)
```

Comment:
I'm not sure if this is where we want to land, changing this from setup_device to the constructor... We can leave it like this for now, but I did have some other thoughts:

I thought maybe we could have "two modes" of opening PCIDevice. When you create a class you'd just be able to read what arch it is, and some other PCI parameters. Then, when calling setup_device, you actually open the device for usage and you get BARs available.

So this would map to constructor not calling any KMD function, and setup_device calling TENSTORRENT_IOCTL_QUERY_MAPPINGS.

These thoughts were during my rewriting experiment, maybe this makes it clearer: https://github.com/tenstorrent/tt-umd/blob/pjanevski/new_api_experiment/ideal_device/pci/pci_device.h#L37

The only argument I have for doing it this way, is that than some other host application (while another is doing whatever with this device) can get arch type or even soc descriptor without mapping all the mappings, just by opening PCIDevice class and reading the arch (so without calling setup_device). While working on Blackhole Buda issues, I ran across some Buda runtime code calling get_arch_type multiple times, which resulted in PCIDevice being opened and closed multiple times (which created some bugs for me). Still not sure if it is a good argument though.

Reply:
The rationale for relying on constructor/destructor rather than separate methods is to provide a stronger guarantee that PCIDevice instances are in a good state.

I haven't carefully considered the problem you describe, and although I intend to leave this change as is for now I am not opposed to refactoring it back in the future if it ends up being the wrong thing.

However, I think improved PCI device enumeration code can address the issue.  (Full topology device enumeration -- what create-ethernet-map does today -- is a different story).  I updated https://github.com/tenstorrent/tt-umd/issues/117 - I agree that we should not be mapping the BAR(s) simply to discover what architecture a chip is.